src/libexpr/print.cc

```diff
@@ -90,5 +127,260 @@ printAttributeName(std::ostream & str, std::string_view name) {
     return str;
 }
 
+typedef std::set<Value *> ValuesSeen;
+
+class Printer
+{
+    private:
+        EvalState & state;
+        PrintOptions options;
```

Comment:
I'm currently passing `PrintOptions` by value everywhere. Should this be a reference instead? I'm not sure what the tradeoffs of copying a couple words around the stack vs. doing a bunch of pointer indirection is.

Reply:
I think it would take a benchmark, but I don't think performance is critical here, so no objections from me.
My reasoning is that printing only happens at the boundary, not within expressions. This puts somewhat of a limit on how much it is used in any given command invocation.